### PR TITLE
Bd token inline hook

### PR DESCRIPTION
**Pull request by brianduffield-okta**
## Description:
- **What's changed?** Adding a new Token Hook Guide

Added a new Guide under the Hooks heading (Hooks > Token Inline Hook).

- **Is this PR related to a Monolith release?** No

## Resolves:
- [OKTA-350189](https://oktainc.atlassian.net/browse/OKTA-350189)

### CODE REVIEW

**Comment by quyle-okta** [comment]
@brianduffield-okta On this page (https://developer.okta.com/docs/reference/token-hook/), can we change the path for Step 1 so it includes the top-level nav menu:

BEFORE: 1. Go to API > Authorization Servers.
AFTER: 1. Go to Security > API > Authorization Servers.

**Comment by brianduffield-okta** [comment]
> @brianduffield-okta On this page (https://developer.okta.com/docs/reference/token-hook/), can we change the path for Step 1 so it includes the top-level nav menu:
> 
> BEFORE: 1. Go to API > Authorization Servers.
> AFTER: 1. Go to Security > API > Authorization Servers.

Good catch. Will update on this branch.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/overview/index.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+title: Overview
+---
+
+The Token Inline Hook can be used to customize the authorization code flow that occurs between an application and the Okta org used for authentication.
+
+This guide provides example code for an external service to respond to calls from a Token Inline Hook, and provides an end-to-end scenario using a local application, an Okta org, and the external service.
+
+### The Scenario
+
+In the following Token Inline Hook scenario, the external service code parses a request from Okta, evaluates the user name against a simple patient data store, and, if the user is a patient, responds to Okta with a command to add a patient ID claim to the token. If the user name is not part of the data store, no action is taken. The token is returned to the local application for authentication.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/overview/index.md:11]: changed to "If the user is not part of the data store"

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+---
+title: Activate and enable
+---
+
+The Token Inline Hook must be activated and enabled within your Okta Admin Console.
+
+- Activating the Token Inline Hook registers the hook with the Okta org and associates it with your external service.
+- Enabling the Token Inline Hook associates the hook with your Okta custom authorization server, which authenticates the Okta-Hosted-Login sample application.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md:8]: This is the title of the application in the repository, so, kept it upper case here. But made sure it matched (Okta-Hosted Login sample), as per: https://github.com/okta/samples-nodejs-express-4

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+---
+title: Extend the sample Express app
+---
+
+The following code extends the local sample Node.js Express application to display the results of the Token Inline Hook claim addition. This step is optional. The Token Inline Hook is functional and the results of the implementation are available in the external service logs, as well the System Log on your Okta org. But this extension is a nice-to-have!
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md:5]: Good catch. Hope the lasts sentence is not too casual :o)

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    - **Authentication field** = `authorization`
+    - **Authentication secret** = `Basic YWRtaW46c3VwZXJzZWNyZXQ=`
+
+6. Click **Save**.
+
+The Token Inline Hook is now set up with a status of active.
+
+### Enable the Token Inline Hook
+
+1. Navigate to **API** > **Authorization Servers**.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md:31]: Yup. Thanks.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+2. Select a Custom Authorization Server from the list (usually **default**).
+
+3. Select the **Access Policies** tab and edit the policy rule to use with the hook. In most cases, edit the Default Policy Rule.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md:35]: I checked, and the "Default Policy Rule" is still applicable ("Default Policy" is the Policy name.) I've added bold font-weight to make this more visible.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+4. From the **Use this Inline Hook** drop-down menu, select the Token Inline Hook you activated ("Patient Token Hook").
+
+5. Click **Update Rule**.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md:39]: Same as previous. Looks like you clicked "Edit" Policy.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/setup-express/index.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+* [https://github.com/okta/samples-nodejs-express-4](https://github.com/okta/samples-nodejs-express-4)
+
+Follow the [README.md](https://github.com/okta/samples-nodejs-express-4/tree/master/okta-hosted-login) instructions to install and run the Okta-Hosted-Login sample application with your Okta org. Make sure to have this application running before proceeding with the Token Inline Hook setup.
+
+To run the sample application:
+
+1. Navigate to your project folder (`samples-nodejs-express-4`).
+
+2. Start your Okta-Hosted-Login server (`npm run Okta-Hosted-Login-Server`).
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/setup-express/index.md:15]: Thanks, removed these steps here as per our discussion.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/setup-express/index.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+---
+title: Set up the sample Express app
+---
+
+The sample Node.js Express application is designed to demonstrate the [Authorization Code flow](/docs/guides/implement-auth-code/overview), and includes an Okta-Hosted-Login server option used in this Token Inline Hook example. Access the code from the following Github repository:
+
+* [https://github.com/okta/samples-nodejs-express-4](https://github.com/okta/samples-nodejs-express-4)
+
+Follow the [README.md](https://github.com/okta/samples-nodejs-express-4/tree/master/okta-hosted-login) instructions to install and run the Okta-Hosted-Login sample application with your Okta org. Make sure to have this application running before proceeding with the Token Inline Hook setup.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/setup-express/index.md:9]: Used title as per the repository ("Okta-Hosted Login sample")

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/overview/index.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+At a high-level, the following workflow occurs:
+
+- A user logs into an Okta-Hosted Login application.
```
[the remark this answers is not in the document]
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/overview/index.md:15]: Title of the application, so, okay. Added "sample" for consistency.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (christinechoy-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md:5]: I thought it was casual and friendly. Is this something we do? If yes, keep it. If not, I'd rephrase it to: This extension is nice to have. Without the "But" and the exclamation mark.

**Review by christinechoy-okta — APPROVED** [review]

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/parse-hook/index.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+---
+title: Parse the Token Inline Hook request
+---
+
+The external service in this scenario requires code to handle the Token Inline Hook request from the Okta org. Use the Glitch example to either build or copy the code (re-mix on Glitch) that parses the Token Inline Hook call.
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/parse-hook/index.md:5]: It doesn't hurt to hyperlink "Glitch example" here again. I did re-mix it in the earlier step but nice to hyperlink it again to confirm that I've re-mixed the right one.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/check-against-store/nodejs/check-patients.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+```JavaScript
+// Example Patients Data Store
+
+const patients = [
+  {
+    username: 'michelle.test@example.com',
+    ExternalServicePatientID: '1235',
+  },
+  {
+    username: 'raj.gupta@example.com',
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/check-against-store/nodejs/check-patients.md:10]: In the Glitch example, the email for Patient ID 6789 is actually "bob.uncle@example.com" instead of "raj.gupta@example.com" - let's make the docs and Glitch code snippets consistent if possible.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+The Token Inline Hook is now set up with a status of active.
+
+### Enable the Token Inline Hook
+
+1. Navigate to **Security** > **API** > **Authorization Servers**.
+
+2. Select a Custom Authorization Server from the list (usually **default**).
+
+3. Select the **Access Policies** tab and edit the policy rule to use with the hook. In most cases, edit the **Default Policy Rule**.
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md:35]: There are actually 2 Edit icon on the Default Policy screen - the top Edit button isn't the one you want. It's the bottom pencil icon but there is no Edit label.  Can we make the instructions more clear on which edit to click on as I naturally gravitate to the top and looking for an "Edit" button and may not be aware that there is also Edit button at the bottom.. 
Screen Shot 2021-02-26 at 4 35 12 PM

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+The following code extends the local sample Node.js Express application to display the results of the Token Inline Hook claim addition. This step is optional. The Token Inline Hook is functional and the results of the implementation are available in the external service logs, as well as the System Log on your Okta org. But this extension is nice to have!
+
+To extend the local sample Node.js Express application, you need to update the `sample-web-server.js` file.
+
+### Update the web server page
+
+1. Navigate to your project folder `samples-nodejs-express-4` and continue to the `common` folder (`samples-nodejs-express-4/common`).
+2. In an editor, open the `sample-web-server.js` page.
+3. Locate the routing function `app.get('/profile'` and add the code, below, after the `const` declarations and before the `res.render` function.
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md:13]: I was a little confused here - we say "add the code below" but the snippet shows the entire function and I only needed to copy the middle section of the code.  Can we either only show the code to be copied or update the instructions to say to modify the routing function as in the example below?

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+2. Start your Okta-Hosted-Login server (`npm run okta-hosted-login-server`).
+
+3. Navigate to your sample application (`http://localhost:8080`).
+
+4. Navigate to your Glitch.com project and open the Console Logs window (**Tools** > **Logs**).
+
+5. Return to your sample application and sign in with an Okta user who is NOT in the Patients data store.
+
+    - The user should sign in as normal; only the user name should appear in the Glitch log window.
+    - If you extended the sample application, click on `Profile` in the left-hand navigation pane. Only the the user info claims are included in the table.
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md:24]: The left navigation appears as "My Profile" instead of "Profile" in my sample app:
Screen Shot 2021-02-26 at 4 46 41 PM

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+6. Sign out of the sample application, and now sign in with an Okta user who IS in the Patients data store.
+
+    - The user should sign in as normal; however, this user should have a patient ID displayed in the Glitch console output, as well as a successful implementation record of the Token Inline Hook, available for review in your Okta org's System Log (**Reports** > **System Log**).
+    - If you extended the sample application, click on `Profile` in the left-hand navigation pane. The patient ID is added as part of the Claims table.
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md:29]: Profile should be "My Profile"

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+3. Navigate to your sample application (`http://localhost:8080`).
+
+4. Navigate to your Glitch.com project and open the Console Logs window (**Tools** > **Logs**).
+
+5. Return to your sample application and sign in with an Okta user who is NOT in the Patients data store.
+
+    - The user should sign in as normal; only the user name should appear in the Glitch log window.
+    - If you extended the sample application, click on `Profile` in the left-hand navigation pane. Only the the user info claims are included in the table.
+
+6. Sign out of the sample application, and now sign in with an Okta user who IS in the Patients data store.
```
Comment (quyle-okta) [inline packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md:26]: I completely missed the NOTE in the "Check data store" section so I have an assumption that others may too - maybe we should put a reminder that the Okta user you're signing in should be part of the data store and also be assigned to your application.

**Comment by quyle-okta** [comment]
@brianduffield-okta one more comment - can you find out how we're able to set the **extPatientId** claim without having to explicitly add it as a custom claim in the guide below?
https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/add-custom-claim/

I thought new and non-standard claims need to be defined first above but I might be missing something...

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/check-against-store/nodejs/check-patients.md (hunk 11 above)
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/check-against-store/nodejs/check-patients.md:10]: thanks, agree. Updated glitch example to match doc.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md (hunk 12 above)
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/activate-and-enable/index.md:35]: Thanks, re-worded that step to orient the user towards the Rule table, i.e.: 

"Select the **Access Policies** tab. Navigate to the Rule table and click the Edit icon next to the policy rule that will use the Inline hook. In most cases, edit the **Default Policy Rule** of the **Default Policy**."

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md (hunk 13 above)
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/extend-express/index.md:13]: I like including the code on either side of the extension as "markers", but understand that it could be confusing. I was wondering about that myself. I'll update the instructions to say "modify" the function as you suggested.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md (hunk 14 above)
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md:24]: Updated. Thanks.

File: packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md (hunk 16 above)
Comment (brianduffield-okta) [reply packages/@okta/vuepress-site/docs/guides/token-inline-hook/test/index.md:26]: Added another reminder at the top of this page. I also included this info directly in the Glitch application as a code comment.

**Comment by brianduffield-okta** [comment]
> @brianduffield-okta one more comment - can you find out how we're able to set the **extPatientId** claim without having to explicitly add it as a custom claim in the guide below?
> https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/add-custom-claim/
> 
> I thought new and non-standard claims need to be defined first above but I might be missing something...

@quyle-okta My assumption is this is by design. From https://developer.okta.com/docs/reference/token-hook/:

"This functionality can be used to add data that is sensitive, calculated at runtime, or complexly-structured and not appropriate for storing in Okta user profiles. Data added this way is never logged or stored by Okta. As an example, tokens minted for a medical app could be augmented with confidential patient data provided by your external service and not stored in Okta."

**Comment by quyle-okta** [comment]
> > @brianduffield-okta one more comment - can you find out how we're able to set the **extPatientId** claim without having to explicitly add it as a custom claim in the guide below?
> > https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/add-custom-claim/
> > I thought new and non-standard claims need to be defined first above but I might be missing something...
> 
> @quyle-okta My assumption is this is by design. From https://developer.okta.com/docs/reference/token-hook/:
> 
> "This functionality can be used to add data that is sensitive, calculated at runtime, or complexly-structured and not appropriate for storing in Okta user profiles. Data added this way is never logged or stored by Okta. As an example, tokens minted for a medical app could be augmented with confidential patient data provided by your external service and not stored in Okta."

Ahh - this makes perfect sense then between a custom claim at runtime vs having a custom claim as part of the standard Okta user profile.  Thanks for the explanation.

**Review by quyle-okta — APPROVED** [review]
